test/extended/ha/ha.go

```diff
@@ -0,0 +1,258 @@
+package ha
+
+/*
+This controllers test suite is not part of the "default" group, because its
+testing involves terminating and launching controllers components of OpenShift.
+Therefore it can't be run in parallel.
+*/
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os/exec"
+	"regexp"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+var _ = g.Describe("ha: Election of OpenShift controllers", func() {
+	defer g.GinkgoRecover()
+	var (
+		podConfig  = exutil.FixturePath("fixtures", "hello-world-pod.json")
+		reExitCode = regexp.MustCompile(`(?i)exitcode:(\d+)`)
+	)
+
+	// Run a controller manager with one intitial controllers instance.
+	// Running controllers are needed for NewCLI to run.
+	// FIXME: seems like OutputDir is unset at this moment (execution of the first test)
+	m, _err := exutil.NewControllersManager(exutil.MasterConfigPath(), 0, 8444, exutil.TestContext.OutputDir)
+	if _err != nil {
+		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
+	}
+	// FIXME: Controllers instance immediatelly terminates after acquiring its lease without following sleep
+	time.Sleep(1 * time.Second)
+
+	// One controllers instance needs to be running before CLI can be initialized
+	_, _err = m.StartNewInstance()
+	if _err != nil {
+		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
+	}
+	_, _, _err = m.WaitForActive(time.Second * 2)
+	if _err != nil {
+		g.Fail(_err.Error())
+	}
+
+	oc := exutil.NewCLI("controller-leases", exutil.KubeConfigPath())
+
+	// WaitForPod blocks until given pod terminates.
+	// FIXME: this is probably too complicated
+	// FIXME: consider using oc logs directly
+	WaitForPod := func(podName string, timeout time.Duration) {
```

Comment:
No, still have to sort through the experience

On Sep 15, 2015, at 12:30 PM, David Eads <notifications@github.com> wrote:

In test/extended/ha/ha.go
<https://github.com/openshift/origin/pull/4628#discussion_r39532364>:

> +	// One controllers instance needs to be running before CLI can be initialized
> +	_, _err = m.StartNewInstance()
> +	if _err != nil {
> +		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
> +	}
> +	_, _, _err = m.WaitForActive(time.Second * 2)
> +	if _err != nil {
> +		g.Fail(_err.Error())
> +	}
> +
> +	oc := exutil.NewCLI("controller-leases", exutil.KubeConfigPath())
> +
> +	// WaitForPod blocks until given pod terminates.
> +	// FIXME: this is probably too complicated
> +	// FIXME: consider using oc logs directly
> +	WaitForPod := func(podName string, timeout time.Duration) {

This is big enough that I'd like to see it factored as a named function.

@smarterclayton <https://github.com/smarterclayton> did your sync command
ever go in?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4628/files#r39532364>.